Add `title_align` command

Aestek opened the pull request:
Controls the window titles alignment in title bars.
Possible values are :
* left
* center
* right

orestisf1993 commented:
This doesn't play nice with marks

Relevant issue: #1750

Airblader commented:
Agreed, it definitely has to work with marks. A potential solution would be to move marks to the left for right-aligned titles and for centered to just make sure that the title ends with an ellipsis when hitting the marks.

Aestek commented:
@orestisf1993 thx for the link. Seeing the issue i'll try to play with `pango_layout_set_alignment()` instead of computing the offset myself.

@Airblader how about keeping a padding (something like `1 * font_height` ?) between the window title and its marks : 
* left align : 
`[title]                                           [padding][marks]`
* center align : 
`[marks_width][padding]   [title]   [padding][marks]`
* right align : 
`[marks][padding]                                           [title]`

This way 
* the window title is always separated from the marks whether there's an ellipsis or not
* in centered mode, the window title is centered even if there is marks

Airblader commented:
> Seeing the issue i'll try to play with pango_layout_set_alignment() instead of computing the offset myself.

The problem is that this doesn't work if pango isn't used. I actually thought it's quite nice that you found a solution that works for both. We could also just use the manual computation in the non-pango fallback.

Not sure I understand the second part. Given a long enough title, we have to use an ellipsis. The main point is that the title should never overlay the marks, but just turn into ellipsis before it bleeds into the marks. We (can) know how wide everything is, so it's just about computing things correctly.

BTW, when you update the PR, screenshots of a few of these cases would be much appreciated. :-)

orestisf1993 commented:
> I actually thought it's quite nice that you found a solution that works for both.

Sorry for the noise but you are the one that found the pango_layout_set_alignment solution: https://github.com/i3/i3/issues/1750#issuecomment-216063705. Or are you referring to something else?

Airblader commented:
Yes, but that one only works for pango. The solution currently in this PR (using text width prediction) can in theory work also for non-pango. :-)

Aestek commented:
I updated the changes to account for the marks. Here what it looks like with a long title and a mark :
I added an extra `logical_px(2)` between mark and title for right alignment because they were too squished together.

left : 
![image](https://user-images.githubusercontent.com/1712219/39147132-18df598a-4739-11e8-8dd5-f83ba37c0752.png)
center : 
![image](https://user-images.githubusercontent.com/1712219/39147180-3b9498be-4739-11e8-93da-cd0c64dbdc6c.png)
right : 
![image](https://user-images.githubusercontent.com/1712219/39147397-c0f50d86-4739-11e8-8d9a-df65901edcc5.png)

What do you guys think ?

Airblader commented:
Haven't looked at the code yet. The pictures look convincing (thanks!), I'd actually prefer a bit more space between title and marks.

What does surprise me just by looking is that the space looks bigger in the centered case than in the left case, and also that in the right case less of the title is shown (logically it should be the same in all cases, no?)

orestisf1993 commented:
How do we feel about a config directive for the default behaviour? I think that `for_window [class=".*"]` rules are not particularly pretty.

orestisf1993 commented on include/data.h:
```diff
@@ -651,6 +651,15 @@ struct Con {
+        ALIGN_LEFT = 0,
+        ALIGN_CENTER = 1,
+        ALIGN_RIGHT = 2
+    } align;
```
I think we can typedef it as `title_align_t` and then …

orestisf1993 commented on src/commands.c:
```diff
@@ -1951,6 +1951,37 @@ void cmd_title_format(I3_CMD, const char *format) {
+    DLOG("setting title_align to \"%s\"\n", direction);
+    HANDLE_EMPTY_MATCH;
+
+    int dir;
```
… and then use the new type for these variables instead of int

orestisf1993 commented on include/data.h:
```diff
@@ -651,6 +651,15 @@ struct Con {
+    } align;
+
+    /** Window's name align */
+    int title_align;
```
This new member is not used here

orestisf1993 commented on src/x.c:
```diff
@@ -589,7 +589,9 @@ void x_draw_decoration(Con *con) {
         goto after_title;
     }
 
+    int title_padding = logical_px(2);
```
let's make this a `const`

orestisf1993 commented on src/x.c:
```diff
@@ -610,9 +612,18 @@ void x_draw_decoration(Con *con) {
             mark_width = predict_text_width(mark);
 
+            int mark_offset_x = 0;
+            switch (con->align) {
```
I think an `if` is simpler here

Airblader replied:
I actually prefer the switch/case. It's an enum, it has a defined set of values, so switch/case is the idiomatic approach.

orestisf1993 replied:
I like switches but in this case it's just an if - else with more keystrokes/lines (and it is missing the `break` in the default case - not needed but generally a good practice). This could even be a ternary operator.

Airblader replied:
I'm effectively OK with either solution here. :-) I guess initially I'd also write it with a ternary.

Airblader commented:
> How do we feel about a config directive for the default behaviour?

I think there's even a case to be made in favor of *only* supporting a global directive, not a per-window command. What's the usecase for aligning titles differently in different windows?

Airblader commented:
What's the status here? Is anyone working on this or should we close it?
